src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,165 @@
+topology). Finally we define the differential of a function in terms of derivations.
+
+The content of this file is not meant to be regarded as an alternative definition to the current
+tangent bundle but rather as a purely algebraic theory that provides a purely algebraic definition
+of the Lie algebra for a Lie group.
+
+-/
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+{M : Type*} [topological_space M] [charted_space H M]
+
+open_locale manifold
+
+namespace point_derivation
+
+instance smooth_functions_algebra : algebra 𝕜 C^∞⟮I, M; 𝕜⟯ := by apply_instance
+instance smooth_functions_tower : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ := by apply_instance
+
+/-- Evaluation at a point is a ring homomorphism. -/
+def smooth_funtion.eval' (x : M) : C^∞⟮I, M; 𝕜⟯ →+* 𝕜 :=
+{ to_fun    := λ f, f x,
+  map_one'  := rfl,
+  map_mul'  := λ f g, rfl,
+  map_zero' := rfl,
+  map_add'  := λ f g, rfl }
+
+/-- The above evaluation gives rise to an algebra structure of `C^∞⟮I, M; 𝕜⟯` on `𝕜`. -/
+def algebra (x : M) : algebra C^∞⟮I, M; 𝕜⟯ 𝕜 := (smooth_funtion.eval' I x).to_algebra
+
+/-- With the above algebra structure evaluation is actually an algebra morphism. -/
+def smooth_function.eval (x : M) :
+  @alg_hom C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ 𝕜 _ _ _ _ (point_derivation.algebra I x) :=
+{ commutes' := λ k, rfl, ..smooth_funtion.eval' I x }
+
+/-- The scalar multiplication defined above gives rise to a module structure. -/
+def module (x : M) : module C^∞⟮I, M; 𝕜⟯ 𝕜 :=
+@algebra.to_module _ _ _ _ (point_derivation.algebra I x)
+
+lemma scalar_def {x : M} {f : C^∞⟮I, M; 𝕜⟯} {k : 𝕜} :
+  @has_scalar.smul C^∞⟮I, M; 𝕜⟯ 𝕜 (point_derivation.algebra I x).to_has_scalar f k = f x * k := rfl
+
+lemma is_scalar_tower (x : M) :
+  @is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ 𝕜 _ (point_derivation.algebra I x).to_has_scalar _ :=
+{ smul_assoc := λ k f h, by { simp only [scalar_def, algebra.id.smul_eq_mul, smooth_map.coe_smul,
+  pi.smul_apply, mul_assoc]} }
+
+end point_derivation
+
+/-- The derivations at a point of a manifold. Some regard this as a possible definition of the
+tangent space -/
+@[reducible] def point_derivation (x : M) :=
+  @derivation 𝕜 C^∞⟮I, M; 𝕜⟯ _ _ _ 𝕜 _ (point_derivation.module I x) _
+    (point_derivation.is_scalar_tower I x)
+
+variable (M)
+
+/-- The total bundle of point derivations. -/
+def derivation_bundle := Σ x : M, point_derivation I x
+
+variables {I M}
+
+/-- The inclusion map of derivations at a point into the total bundle. -/
+def derivation_inclusion {x : M} (v : point_derivation I x) : derivation_bundle I M :=
+sigma.mk x v
+
+instance [inhabited M] : inhabited (derivation_bundle I M) :=
+⟨derivation_inclusion (0 : point_derivation I (default M))⟩
+
+section
+
+/- Why do I need to rewrite extensionality rules for reducible defs? -/
```

Comment:
What goes wrong? You don't even need to write extensionality rules for semireducible defs, the `ext` tactic is very aggressive!

Reply:
Actually nothing, I wrote this one year ago and I don't know what went wrong anymore...